### PR TITLE
Tutorial & updates to the DynamicBatchSampler

### PR DESCRIPTION
**Pull request by vitas-ai**
Hi, 

I added a method to reduce unused padding frames, and an alternative computation of the bucket boundaries.
Please take a look at the logging messages if they are fine. The validation set also needs a DynamicBatchSampler on low-VRAM.

Please compare with the shared tutorial draft.

Best wishes,
Andreas

### CODE REVIEW

**Comment by mravanelli** [comment]
Hi @vitas-ai,
thank you for this PR. Here are a couple of comments:
1. We don't want to change the librispeech recipe directly. Dynamic batching usually works well, but it introduces some parameters that might be a bit obscure to the users (e.g., those for bucketing). Moreover, if we do it for one recipe we should do it for all the others to be consistent (this is something we might want to consider for the future). Instead, creating a tutorial that explain how to use it is totally fine for the moment. My proposal is that, instead of changing the official recipes, we can create one more template (e.g, `templates/speech_recognition/ASR/train_with_dynamic_batching.py`) that shows a working code for doing ASR with dynamic batching. Users can take a look at that if they want to use it. 
2. There are some tests that are failing (both linters and actual tests). Could you please take a look at that?

**Comment by vitas-ai** [comment]
Hi Mirco,

I took it from the other PR -> original recipes are restored; I made some fixes & hope the tests are working out now.

I'll get to a template later, when we have more clarity here. Hope that's ok :)

Best wishes,
Andreas

**Comment by mravanelli** [comment]
@vitas-ai any news on that? As for the tutorial I know @popcornell  is working on that as well. @popcornell any update on your side?

**Comment by vitas-ai** [comment]
Hi @mravanelli, I will deal with the checks - other errors keep popping up.

For the tutorial, @popcornell wanted to do a few changes. I stopped then.

What both of us saw: I continued with a 6+ months old PR, and there seem to be some issues for everything running smoothly. Should I merge the latest dev status in the `dynamic_libri` branch? (see this as the only solution right now, so both of us can continue faster with the tutorial).

**Comment by mravanelli** [comment]
Yes, I think it is better to work with the latest dev. @popcornell any update on your side on the tutorial?

**Comment by vitas-ai** [comment]
@popcornell I saw that TIMIT recipes used dynamic batching, I put them into separate files, so that the original `train.py` and its original hparams file are preserved. Yet, I did not test them further.

@mravanelli putting dev back in ... it's a lot of files that github marked as 'changed' - sorry for that. I hope it clarifies once the merging process is started, that there are just a few changes.

crossing fingers all these tests work out this time...

**Comment by TParcollet** [comment]
Oh alright I understand! it's because this is not a pull request for merging to develop but to an other old branch !!

**Comment by popcornell** [comment]
> I saw that TIMIT recipes used dynamic batching, I put them into separate files, so that the original train.py and its original hparams file are preserved. Yet, I did not test them further.

we can probably have just one file with some option like use_dynamic_batching. Let me merge them. 
I ll complete the tutorial by tomorrow. Sorry I was not that well in these last couple of weeks.

**Comment by mravanelli** [comment]
As mentioned in the meeting, it looks like this PR contains things not related to the dynamic batching part. @popcornell could you please clean it up and push your changes?

**Comment by vitas-ai** [comment]
Sorry, the files `recipes/CommonVoice/ASR/seq2seq/hparams/train_de.yaml` and `recipes/CommonVoice/ASR/seq2seq/README.md` slipped in - that was not on intention.

They are removed now.

**Comment by popcornell** [comment]
I have made a PR to your branch. 
I am highlighting here some changes I am proposing.

I think we can get rid of `left_bucket_len` and multiplier and just use your warping approach:
https://github.com/popcornell/speechbrain/blob/b76caae391171a1ad8791d34a3f3bc98f5003854/speechbrain/dataio/sampler.py#L415

Also I added another arg `max_batch_ex`: https://github.com/popcornell/speechbrain/blob/b76caae391171a1ad8791d34a3f3bc98f5003854/speechbrain/dataio/sampler.py#L419

We want sometimes to prevent that batches have too many examples.

File: recipes/LibriSpeech/ASR/seq2seq/hparams/train_BPE_1000_dynamic_batching.yaml [hunk 1]
```diff
@@ -0,0 +1,298 @@
+# ############################################################################
```
Comment (popcornell) [inline recipes/LibriSpeech/ASR/seq2seq/hparams/train_BPE_1000_dynamic_batching.yaml:1]: I have merged this file with the standard one.

File: recipes/TIMIT/ASR/seq2seq/hparams/train_dynamic_batching.yaml [hunk 2]
```diff
@@ -0,0 +1,217 @@
+# ################################
```
Comment (popcornell) [inline recipes/TIMIT/ASR/seq2seq/hparams/train_dynamic_batching.yaml:1]: Also this one merged, same for the training scripts.

File: speechbrain/core.py [hunk 3]
```diff
@@ -719,12 +720,13 @@ def _train_loader_specifics(self, dataset, loader_kwargs):
                 self.train_sampler = DistributedSamplerWrapper(
                     sampler,
                     rank=self.rank,
-                    drop_last=drop_last,
-                    shuffle=shuffle,
+                    # drop_last=drop_last,
```
Comment (popcornell) [inline speechbrain/core.py:723]: I understand why shuffle is dropped. But what about drop_last ?

Comment (vitas-ai) [reply speechbrain/core.py:723]: no clue. I think it was this way somewhere? not on purpose...

**Comment by mravanelli** [comment]
@vitas-ai, at which stage we are here? It looks like you and @popcornell are close to converge on a final PR, right?

**Comment by vitas-ai** [comment]
Yes. We are just sorting out for how to proceed. The legacy dev branch looms to come back in again - would like to avoid that, @popcornell and I are writing on Slack to resolve this.

**Comment by popcornell** [comment]
Closing, moved to https://github.com/speechbrain/speechbrain/pull/1174
